### ISSUE TITLE
Brazilian Portuguese translation

### ISSUE THREAD

**Issue by donnemartin** [donnemartin/system-design-primer#40]
Maintainer(s): @IuryAlves

Please check out the [Translations Contributing Guidelines](https://github.com/donnemartin/system-design-primer/blob/master/CONTRIBUTING.md#translations).

Interested in helping?  Let us know!

**Comment by hbarcelos** [comment]
Hello, I'm interested in helping.

What can I do next?

**Comment by IuryAlves** [comment]
Hi @hbarcelos. The maintainer of this project is not available, see issue #66. So i think we could work in my [fork](https://github.com/IuryAlves/system-design-primer). When the translation is done we send a pull request to this repo.

We are going to do the translation in the file called`README-pt-BR.md`. Also you can translate whole sections or just some sentences, it`s your choice.

**Comment by heltucosta** [comment]
I would like to help also.

**Comment by SuperAssassinNinja** [comment]
Hey guys, 
I'd like to help, how do we proceed?

**Comment by IuryAlves** [comment]
Hi @heltucosta @SuperAssassinNinja. What do you think about my suggestion? If you have better ideias, please let me know.

**Comment by GSalomaoK** [comment]
Sounds good to me. I'm able to help also.

**Comment by IuryAlves** [comment]
I have created [issues](https://github.com/IuryAlves/system-design-primer/issues) to the brazilian translation. Each issue is a section of the text. Just make a comment in the issue that you want to translate.

**Comment by heltucosta** [comment]
Sounds good to me. will take a look

**Comment by donnemartin** [comment]
Hi, I'm back now, sorry for the delay.

Great to see the activity on this, thanks for the help!  Let me know if you need anything.

**Comment by IuryAlves** [comment]
I am struggling to find a good translation to the word `worker` in the [Application Layer Section](https://github.com/donnemartin/system-design-primer#application-layer)

Should i keep the original word `(worker)` ?
Should i try to explain what a `(worker)` is ?

cc @donnemartin, @hbarcelos, @heltucosta, @SuperAssassinNinja , @GSalomaoK

**Comment by hbarcelos** [comment]
I think worker is a well known jargon in Software Engineering and has no perfect translation.
Also, since this repo targets more experienced engineers that want to become system architects, I believe there's no need to define what a worker is.

What the rest of you think?

**Comment by ricardovsilva** [comment]
@hbarcelos, this content targets more experienced engineers but it can also be very very useful to rookies developers. My sugestion is to just hyperlink word worker to some url with good content about. For example: https://developers.google.com/web/fundamentals/getting-started/primers/service-workers?hl=pt-br

**Comment by donnemartin** [comment]
I think these are good suggestions, thanks @hbarcelos and @ricardovsilva!

**Comment by IuryAlves** [comment]
Hi @EarlyChannelBR i was translating on my fork [https://github.com/IuryAlves/system-design-primer](https://github.com/IuryAlves/system-design-primer).

There is a PR for a partial translation: https://github.com/donnemartin/system-design-primer/pull/67/files

**Comment by ricardovsilva** [comment]
@IuryAlves I'm going to review this PR no later than this weekend.

**Comment by IuryAlves** [comment]
Thank you @ricardovsilva

**Comment by vicksbr** [comment]
Hello! I want to help in this project so just let me know what i can do!

**Comment by IuryAlves** [comment]
@vicksbr 

I haven't worked in the translation for a long time.

Maybe I need to hand it over to someone else.

cc @donnemartin

**Comment by AlexandreCarvalhoMelo** [comment]
Hey i'd like to keep on working on the Translation, how can we?

**Comment by IuryAlves** [comment]
@vicksbr @AlexandreCarvalhoMelo 

I have done a fork of this project for the PT-br translation.

https://github.com/IuryAlves/system-design-primer

You can pick up one of the open tickets.

**Comment by donnemartin** [comment]
Hi @IuryAlves : Are you still interested in translating the guide and being the language maintainer?

**Comment by IuryAlves** [comment]
Hi @donnemartin 

I am not active maintaining it. All the translation that I did is [here](https://github.com/IuryAlves/system-design-primer)

**Comment by slompo** [comment]
> Hi @donnemartin
> 
> I am not active maintaining it. All the translation that I did is [here](https://github.com/IuryAlves/system-design-primer)

The main repository is already updated?

**Comment by IuryAlves** [comment]
@slompo 

There is a PR open: https://github.com/donnemartin/system-design-primer/pull/67

**Comment by wilkmaia** [comment]
@donnemartin I can help maintaining the Brazilian Portuguese translation. What should the next steps be? Should we keep 
 working on IryAlves' fork and #67?

**Comment by brcavallaro** [comment]
As @wilkmaia, I am interested in contributing as well, @donnemartin. Looking forward for next steps here :)

**Comment by hugosantos-io** [comment]
Hi @donnemartin! Together with @wilkmaia and @brcavallaro I'm also available to help maintain the pt-BR version.

**Comment by caiocinel** [comment]
I'm also available to help with the translation!

**Comment by IuryAlves** [comment]
It is not active. Feel free to fork this repo and translate it.

**Comment by geovanasilva** [comment]
Hello @donnemartin, I find out that the pt-br maintainer is not longer active working in the translations so I took the liberty to contribute using my [fork](https://github.com/geovanasilva/system-design-primer/tree/docs/pt-br-translations), special thanks for @kevenLeandro who gave me the idea and will also be working with me on the translations. I also have a question: the repository has a [PR ](https://github.com/donnemartin/system-design-primer/pull/67) opened about the creation of pt-BR README, should I refer my fork on it, open another PR or just open the request once I finish the whole translation?

**Comment by valdemarjuniorr** [comment]
I would like to help, but I don't know how. Anyone can help to start? Do I need to fork the repo and translate it? Thanks in advance.

**Comment by diegocmsantos** [comment]
Nice initiative.
I can help with it as well.

**Comment by geovanasilva** [comment]
@valdemarjuniorr, @diegocmsantos: I'm waiting for permission from @donnemartin, but feel free to make contributions to my fork and at the end I'll open PR for the final translation 🥳

**Comment by leonardodecuritiba** [comment]
> @valdemarjuniorr, @diegocmsantos: I'm waiting for permission from @donnemartin, but feel free to make contributions to my [fork](https://github.com/geovanasilva/system-design-primer/tree/docs/pt-br-translations) and at the end I'll open PR for the final translation. What do you guys thinks? 🤔

Hi guys, I want to contribute with this repo.
What I need to do?
@geovanasilva can you help me?

**Comment by avulso17** [comment]
I want to help!
How can I contribute for this issue?

**Comment by franciscocxavier** [comment]
# I really wanted to contribute to this issue
Hey, guys. I really wanted to contribute to this issue. However, please note that this thread has not been updated for many days. I will work on this problem in my fork, please if you have any information reply or tell me where to look for resources

**Comment by geovanasilva** [comment]
Hey everyone! I've created a [fork](https://github.com/geovanasilva/system-design-primer/tree/docs/pt-br-translations) and I'd love for us to collaborate on it. @franciscocxavier feel free to open pull requests so we can work on it together. If you'd like to chat about it, just let me know and we can set up a meeting. Looking forward to working together!

**Comment by leonardodecuritiba** [comment]
Hello Geovana, how are you?
Can I schedule a meet with you?
I wanted to contribute with this translation.
Can you teach me, on how to get the things done?

Thanks.,

Leonardo Zanin

**Comment by geovanasilva** [comment]
Hello, I am fine and you? @leonardodecuritiba. I've been attempting to contact the owner of the repository since 2023, but I have not received any responses. Therefore, I updated the fork I had created and incorporated some changes from the first pull request opened by @IuryAlves, I've [opened a new pull request](https://github.com/donnemartin/system-design-primer/pull/867) with these changes, and now I need the repository owners to actively review the repository for approval. However, they may be no longer active, as the last merge into the repository was a year ago. I also tried to contact @donnemartin via email to check out. I will keep you guys informed on any updates. At the moment, I don't think it makes sense to meet for collaboration due to this obstacle, but thank you all for your commitment.

**Comment by BrunoCiccarino** [comment]
I'm Brazilian, if you want, I'll translate!

**Comment by geovanasilva** [comment]
Hello, I hope everyone is doing well. I would like your support in [this review](https://github.com/donnemartin/system-design-primer/pull/67), in the comment I mention what is being translated. For anyone interested in collaborating on the translation, please send me a friend request at Discord my username **@morticcian**. I've created a community on a Discord server. Currently it looks like the personalized invite links to the server are only available for premium users. However, there's a good limit on the number of members per server, so I think that will be sufficient. I will add the members manually to the server so I count on everyone's understanding but we can improve the community and this process later. Thank you 💜

**Comment by geovanasilva** [comment]
Hello, I hope everyone is doing well. I would like your support in [this review](https://github.com/donnemartin/system-design-primer/pull/67), in the comment I mention what is being translated. For anyone interested in collaborating on the translation, please enter this community at [Discord](https://discord.gg/66G8YQENQd). Thank you 💜

**Comment by franciscocxavier** [comment]
Hi Geovana,

I apologize for being away from the topic for a while. I've joined the Discord community and am available to collaborate on the translation.

Thank you and let's move forward! 💜

**Comment by biangomes** [comment]
Hey everybody, how can I help?

**Comment by rodri-oliveira-dev** [comment]
Hello everyone,

My name is Rodrigo, and I am a Brazilian Portuguese native speaker working as a Software Architect.

I would like to resume the Brazilian Portuguese translation of the System Design Primer and volunteer as an active maintainer for the pt-BR version.

I have created a fork to organize the work:

https://github.com/rodri-oliveira-dev/system-design-primer-pt-br

My intention is to:

* Translate the current English `README.md` into `README-pt-BR.md`;
* Review and reuse previous translation work where it is still applicable;
* Keep the translation technically accurate and synchronized with the upstream repository;
* Define a consistent glossary for system design terminology in Brazilian Portuguese;
* Invite other native speakers to review and contribute;
* Maintain the translation after it is merged.

I am aware of the previous work in PR #67 and PR #867, and I intend to preserve relevant contributions and credits rather than restart the work without considering what has already been done.

Before proceeding further, I would like to confirm whether this is still the recommended workflow and whether I may take responsibility for maintaining the Brazilian Portuguese translation.

Thank you!

**Comment by salamat78vd** [comment]
Thanks for volunteering to take this on.

That sounds like a well-organized approach. I appreciate the focus on
preserving previous contributions and keeping the translation synchronized
with the upstream project. Looking forward to seeing the progress.

**Comment by boua8745s12** [comment]
Thanks for the encouragement.

I'm looking forward to seeing the translation effort move forward. Keeping
it synchronized with the upstream project while preserving previous
contributions will be a great benefit to the community.

**Comment by rodri-oliveira-dev** [comment]
Thank you for the encouragement!

I’ll move forward with the translation in my fork. My first steps will be:

* Synchronize the fork with the current upstream repository;
* Create the `README-pt-BR.md` file based on the latest English `README.md`;
* Review the previous work from PRs #67 and #867;
* Define a consistent glossary for technical terminology;
* Organize the translation into smaller sections for easier review and collaboration.

I’ll post relevant progress updates here and invite Brazilian Portuguese native speakers to review the translation.

Once the translation is complete, synchronized with the English version, and reviewed, I’ll open a pull request to the original repository.

Contributions and technical reviews are welcome through the fork:

https://github.com/rodri-oliveira-dev/system-design-primer-pt-br

Thank you!

**Comment by rodri-oliveira-dev** [comment]
Hi everyone,

I’ve started working on the Brazilian Portuguese translation.

The current work is available here:

* Translation branch: https://github.com/rodri-oliveira-dev/system-design-primer-pt-br/tree/docs/pt-br-translation
* Draft pull request: https://github.com/rodri-oliveira-dev/system-design-primer-pt-br/pull/1
* Contribution guidelines in Portuguese: https://github.com/rodri-oliveira-dev/system-design-primer-pt-br/blob/docs/pt-br-translation/CONTRIBUTING-pt-BR.md

So far, the translation includes the introduction, study guide, system design interview approach, and the initial sections on scalability, performance, consistency, and availability.

The translation is still in progress. Anyone interested in helping with translation, technical review, language review, links, or terminology is welcome to contribute through the fork.

Thank you!

**Comment by rodri-oliveira-dev** [comment]
Hello everyone,

The Brazilian Portuguese translation is now available for review in draft pull request #1350:

https://github.com/donnemartin/system-design-primer/pull/1350

The main translation is complete and includes:

- `README-pt-BR.md`, with the Brazilian Portuguese version of the guide;
- `CONTRIBUTING-pt-BR.md`, with contribution guidelines and a terminology glossary;
- a small update to `CONTRIBUTING.md` linking to the pt-BR contribution guide.

The translation was compared with the current English `README.md` and went through structural review, terminology consistency review, two assisted technical and linguistic review passes, internal anchor validation, GitHub Flavored Markdown rendering checks, and an external link audit.

The pull request is currently marked as a draft because I would still appreciate an independent review from native Brazilian Portuguese speakers, especially regarding:

- technical accuracy;
- naturalness of the language;
- consistency of system design terminology;
- fidelity to the English source.

I would also like to thank everyone who contributed to the previous translation efforts, especially the work discussed in #67 and #867.

I am available to maintain the Brazilian Portuguese translation and keep it synchronized with future changes to the English version.

Reviews and feedback are very welcome. Thank you!

**Comment by wafir645fcqs7** [comment]
Thanks for taking this on! The review process sounds very thorough.
Looking forward to seeing the community's feedback on the translation.

**Comment by yulinlina** [comment]
**Approach:** Add a Brazilian Portuguese translation stub that points to the original README and translation contribution guidelines.

**Commit message:** `docs: add Brazilian Portuguese translation stub`

**Changes:**

- `README-pt-BR.md` (create)

**Comment by rodri-oliveira-dev** [comment]
Thanks for the suggestion!

PR #1350 already contains the complete Brazilian Portuguese translation, so I kept the full `README-pt-BR.md` instead of creating a translation stub.

However, your comment helped identify a valid navigation issue: the main `README.md` was still linking “Português do Brasil” to issue #40. I have now updated the pull request so that the language selector points directly to `README-pt-BR.md`.

Issue #40 will remain referenced for coordination and historical context.

Thank you for pointing this out!

**Comment by yulinlina** [comment]
Agreed — with #1350 already containing the full translation, a stub would now be redundant. The useful fix is the navigation issue you mentioned: the language selector should point directly to the translated file.

Something like this in the main `README.md`:

```diff
- [Português do Brasil](https://github.com/donnemartin/system-design-primer/issues/40)
+ [Português do Brasil](README-pt-BR.md)
```

A few checks that may be useful before marking the PR ready:

- Verify internal anchors in `README-pt-BR.md` resolve correctly after heading translation, since GitHub generates anchors from the rendered heading text.
- Ensure links to `CONTRIBUTING-pt-BR.md` are relative and render correctly on GitHub.
- Add a short note in `README-pt-BR.md` that it tracks the English `README.md` and that translation issues/updates should go through the issue/PR process.
- Include the upstream English commit SHA in the PR description, which makes future syncs easier to audit.

Keeping issue #40 as the coordination/historical thread makes sense. If helpful, I can do a focused pass on link integrity, anchors, and Markdown rendering after the latest update.

---
*via [GitHub Trend-Hacking Pipeline](https://github.com/yulinlina/start_rise)*

**Comment by rodri-oliveira-dev** [comment]
Thanks for the follow-up!

The language selector has already been updated in PR #1350 to point directly to `README-pt-BR.md`.

The other suggested checks have also been covered:

- internal anchors were validated against the translated headings;
- links to `CONTRIBUTING-pt-BR.md` use relative paths;
- GitHub Flavored Markdown rendering was validated;
- `README-pt-BR.md` already includes a note explaining that it tracks the English `README.md` and that contributions are welcome through pull requests.

I’ll also add the English source baseline commit to the pull request description to make future synchronization easier to audit.

Thank you!

**Comment by yulinlina** [comment]
Great — with #1350 containing the full translation, the stub is unnecessary. The main remaining improvement is making future syncs auditable.

For the baseline, I’d put the full upstream SHA in the PR description rather than a branch name:

```markdown
Upstream translation baseline: donnemartin/system-design-primer@<full-sha>
```

You can generate it with:

```bash
git remote add upstream https://github.com/donnemartin/system-design-primer.git
git fetch upstream
git rev-parse upstream/master
```

If you want the reference to survive after merge, consider adding a small maintenance note in `CONTRIBUTING-pt-BR.md`:

```markdown
## Sync baseline

This translation tracks the English `README.md` from upstream commit `<full-sha>`.
```

Then future synchronization can be checked with:

```bash
git diff <full-sha>..upstream/master -- README.md CONTRIBUTING.md
```

One extra thing worth a quick manual check: anchors for translated headings with accents can render differently depending on whether the link uses the raw Unicode fragment or the URL-encoded form, so clicking a few TOC links from the rendered GitHub page is still the safest test.

Once the baseline SHA is added, marking the PR ready and asking maintainers/native reviewers for a final pass seems like the right next step. I can also do a focused review of links, anchors, and rendering on the updated PR.

**Comment by rodri-oliveira-dev** [comment]
> Great — with [#1350](https://github.com/donnemartin/system-design-primer/pull/1350) containing the full translation, the stub is unnecessary. The main remaining improvement is making future syncs auditable.
> 
> For the baseline, I’d put the full upstream SHA in the PR description rather than a branch name:
> 
> Upstream translation baseline: donnemartin/system-design-primer@<full-sha>
> You can generate it with:
> 
> git remote add upstream https://github.com/donnemartin/system-design-primer.git
> git fetch upstream
> git rev-parse upstream/master
> If you want the reference to survive after merge, consider adding a small maintenance note in `CONTRIBUTING-pt-BR.md`:
> 
> ## Sync baseline
> 
> This translation tracks the English `README.md` from upstream commit `<full-sha>`.
> Then future synchronization can be checked with:
> 
> git diff <full-sha>..upstream/master -- README.md CONTRIBUTING.md
> One extra thing worth a quick manual check: anchors for translated headings with accents can render differently depending on whether the link uses the raw Unicode fragment or the URL-encoded form, so clicking a few TOC links from the rendered GitHub page is still the safest test.
> 
> Once the baseline SHA is added, marking the PR ready and asking maintainers/native reviewers for a final pass seems like the right next step. I can also do a focused review of links, anchors, and rendering on the updated PR.

Thank you for the suggestion!

The pull request description now records the full upstream baseline:

`donnemartin/system-design-primer@ae9bbd7b02d90b9866215de185217d33f39ab733`

This is the exact upstream commit used as the source for the translation and should make future synchronization changes easier to audit.
